File: benchmark-csharp/src/main/csharp/RecursiveTest.cs
using System;
using System.Collections.Generic;
using System.Text;

/**
 * RecursiveTest.
 *
 * @author Alexandre
 *
 */
sealed class RecursiveTest
{
    /**
     * NB_TESTS
     */
    private static int NB_TESTS = 5;

    /**
     * NB_OF_EXCLUSION_MIN_MAX
     */
    private static int NB_OF_EXCLUSION_MIN_MAX = 2;

    /**
     * NB_RECURSIVE_TESTS
     */
    private static int NB_RECURSIVE_TESTS = 1;
    
    /**
     * FIBO_N
     */
    private static int FIBO_N = 41;
    /**
     * averageTimeWithoutMinMax.
     *
     * @param executionTimes
     * @param numberOfMinMaxToRemove
     * @return
     */
    public static long averageTimeWithoutMinMax(List<long> executionTimes, int numberOfMinMaxToRemove)
    {
        long average = 0;
		int minMax = (numberOfMinMaxToRemove > 0) ? numberOfMinMaxToRemove : 0;
		if (minMax != 0) 
		{
        	executionTimes.Sort();
		}
        for (int i = minMax; i < executionTimes.Count - minMax; i++)
        {
            average += executionTimes[i];
        }
        average = average / (executionTimes.Count - minMax * 2);
        return average;
    }


    /**
     * relativeDeviationTimeWithoutMinMax.
     *
     * @param executionTimes
     * @param numberOfMinMaxToRemove
     * @return
     */
    public static double relativeDeviationTimeWithoutMinMax(List<long> executionTimes, int numberOfMinMaxToRemove)
    {
        long averageTimeWithoutMinMaxVariable = averageTimeWithoutMinMax(executionTimes, numberOfMinMaxToRemove);
        long deviation = 0;
		int minMax = (numberOfMinMaxToRemove > 0) ? numberOfMinMaxToRemove : 0;
		if (minMax != 0) 
		{
        	executionTimes.Sort();
		}
        for (int i = minMax; i < executionTimes.Count - minMax; i++)
        {
            deviation += (long)Math.Pow(executionTimes[i] - averageTimeWithoutMinMaxVariable, 2);
        }
        return ((int)((Math.Sqrt(deviation / (executionTimes.Count - 2 * minMax)) / averageTimeWithoutMinMaxVariable * 100) * 100)) / 100d;
    }

    /**
     * testFibo1.
     *
     * @return
     */
    public static long testFiboStatic()
    {
        int fibo = 0;
	    DateTime start = DateTime.Now;
	    for (int i = NB_RECURSIVE_TESTS; i!= 0; i--)
	    {
	        fibo = FiboInstance.fibStatic(FIBO_N);
	    }
	    DateTime end = DateTime.Now;
	    TimeSpan executionTime = end - start;
        Console.WriteLine("[RecursiveTest], Invoke of " + NB_RECURSIVE_TESTS + " fibo = FiboInstance.fibStatic(" + FIBO_N + ")  , fibo=" + fibo + " , snapshot time,"
		    + executionTime.TotalMilliseconds);
	    return (long) executionTime.TotalMilliseconds;
    }
    

    /**
 * testFibo2.
 *
 * @return
 */
    public static long testFiboInstance()
    {
        FiboInstance fiboInstance = new FiboInstance();
        int fibo = 0;
        DateTime start = DateTime.Now;
        for (int i = NB_RECURSIVE_TESTS; i != 0; i--)
        {
            fibo = fiboInstance.fib(FIBO_N);
        }
        DateTime end = DateTime.Now;
        TimeSpan executionTime = end - start;
        Console.WriteLine("[RecursiveTest], Invoke of " + NB_RECURSIVE_TESTS + " fibo = fiboInstance.fib(" + FIBO_N + ")  , fibo=" + fibo + ", snapshot time,"
            + executionTime.TotalMilliseconds);
        return (long)executionTime.TotalMilliseconds;
    }



    /**
    * testFibo2.
    *
    * @return
    */
    public static long testFiboInstanceImplWithInterface()
    {
        Fibo fibo = new FiboImpl();
        int fiboResult = 0;
        DateTime start = DateTime.Now;
        for (int i = NB_RECURSIVE_TESTS; i != 0; i--)
        {
            fiboResult = fibo.fib(FIBO_N);
        }
        DateTime end = DateTime.Now;
        TimeSpan executionTime = end - start;
        Console.WriteLine("[RecursiveTest], Invoke of " + NB_RECURSIVE_TESTS + " fiboResult = fibo.fib(" + FIBO_N + ")  , fibo=" + fiboResult + ", snapshot time,"
            + executionTime.TotalMilliseconds);
        return (long)executionTime.TotalMilliseconds;
    }



    /**
* testFibo2.
*
* @return
*/
    public static long testFiboInstanceImplWithoutInterface()
    {
        FiboImpl fiboImpl = new FiboImpl();
        int fibo = 0;
        DateTime start = DateTime.Now;
        for (int i = NB_RECURSIVE_TESTS; i != 0; i--)
        {
            fibo = fiboImpl.fib(FIBO_N);
        }
        DateTime end = DateTime.Now;
        TimeSpan executionTime = end - start;
        Console.WriteLine("[RecursiveTest], Invoke of " + NB_RECURSIVE_TESTS + " fibo = fiboImpl.fib(" + FIBO_N + ")  , fibo=" + fibo + ", snapshot time,"
            + executionTime.TotalMilliseconds);
        return (long)executionTime.TotalMilliseconds;
    }


        /**
     * main.
     *
     * @param args
     */
    public static void main(string[] args)
    {
	    int nbTests = (args != null && args.Length >= 1) ? int.Parse(args[0]) : NB_TESTS;
	    int nbOfExclusionMinMax = (args != null && args.Length >= 2) ? int.Parse(args[1]) : NB_OF_EXCLUSION_MIN_MAX;

	    List<long> executionTimes = new List<long>(nbTests);

	    for (int i = nbTests; i != 0; i--)
            executionTimes.Add(RecursiveTest.testFiboStatic());
        executionTimes.Sort();
        Console.WriteLine("[RecursiveTest], Invoke of " + NB_RECURSIVE_TESTS + " fibo = FiboInstance.fib(" + FIBO_N + ")  ,, average time,"
                + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) + ", min time," + executionTimes[0] + ", max time," + executionTimes[executionTimes.Count - 1]
                + ", relative deviation time," + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));
	    executionTimes.Clear();
    	
	    for (int i = nbTests; i != 0; i--)
	        executionTimes.Add(RecursiveTest.testFiboInstance());
        executionTimes.Sort();
        Console.WriteLine("[RecursiveTest], Invoke of " + NB_RECURSIVE_TESTS + " fibo = fiboInstance.fib(" + FIBO_N + ")  ,, average time,"
                + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) + ", min time," + executionTimes[0] + ", max time," + executionTimes[executionTimes.Count - 1]
                + ", relative deviation time," + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));
	    executionTimes.Clear();

        for (int i = nbTests; i != 0; i--)
            executionTimes.Add(RecursiveTest.testFiboInstanceImplWithInterface());
        executionTimes.Sort();
        Console.WriteLine("[RecursiveTest], Invoke of " + NB_RECURSIVE_TESTS + " fiboResult = fibo.fib(" + FIBO_N + ")  ,, average time,"
                + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) + ", min time," + executionTimes[0] + ", max time," + executionTimes[executionTimes.Count - 1]
                + ", relative deviation time," + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));
        executionTimes.Clear();

        for (int i = nbTests; i != 0; i--)
            executionTimes.Add(RecursiveTest.testFiboInstanceImplWithoutInterface());
        executionTimes.Sort();
        Console.WriteLine("[RecursiveTest], Invoke of " + NB_RECURSIVE_TESTS + " fibo = fiboImpl.fib(" + FIBO_N + ")  ,, average time,"
                + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) + ", min time," + executionTimes[0] + ", max time," + executionTimes[executionTimes.Count - 1]
                + ", relative deviation time," + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));
        executionTimes.Clear();
	
    }
    
    
    
}

/**
 * Fibo.
 *
 * @author Alexandre
 *
 */
sealed class FiboInstance
{

    /**
     * fib.
     *
     * @param n
     * @return
     */
    public static int fibStatic(int n)
    {
	    if (n < 2)
        //if (n == 0 || n == 1)
	        return (1);
        return (fibStatic(n - 2) + fibStatic(n - 1));
    }

    /**
 * fib.
 *
 * @param n
 * @return
 */
    public int fib(int n)
    {
        if (n < 2)
        //if (n == 0 || n == 1)
            return (1);
        return (fib(n - 2) + fib(n - 1));
    }
}

interface Fibo
{
    int fib(int n);
}

/**
 * Fibo.
 *
 * @author Alexandre
 *
 */
sealed class FiboImpl : Fibo
{

/**
 * fib.
 *
 * @param n
 * @return
 */
    public int fib(int n)
    {
        if (n < 2)
        //if (n == 0 || n == 1)
            return (1);
        return (fib(n - 2) + fib(n - 1));
    }
}

File: benchmark-java/src/main/java/RecursiveTest.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;


/**
 * RecursiveTest.
 * 
 * @author anavarro122404 - 25 sept. 07
 *
 *
 * <!-- $Id: RecursiveTest.java,v 1.2 2007/11/20 13:17:31 anavarro Exp $ -->.
 *
 */
public final class RecursiveTest
{
    /**
     * NB_TESTS
     */
    private static final int NB_TESTS                = 5;

    /**
     * NB_OF_EXCLUSION_MIN_MAX
     */
    private static final int NB_OF_EXCLUSION_MIN_MAX = 2;

    /**
     * NB_RECURSIVE_TESTS
     */
    private static final int NB_RECURSIVE_TESTS      = 1;

    /**
     * FIBO_N
     */
    private static final int FIBO_N                  = 41;

    /**
     * averageTimeWithoutMinMax.
     * 
     * @param executionTimes
     * @param numberOfMinMaxToRemove
     * @return
     */
    @SuppressWarnings("boxing")
    public static long averageTimeWithoutMinMax(final List<Long> executionTimes, final int numberOfMinMaxToRemove)
    {
        long average = 0;
        final int minMax = (numberOfMinMaxToRemove > 0) ? numberOfMinMaxToRemove : 0; 
        if (minMax != 0)
        {
            Collections.sort(executionTimes);
        }
        for (int i = minMax; i < executionTimes.size() - minMax; i++)
        {
            average += executionTimes.get(i);
        }
        average = average / (executionTimes.size() - minMax * 2);
        return average;
    }
    
    
    /**
     * relativeDeviationTimeWithoutMinMax.
     *
     * @param executionTimes
     * @param numberOfMinMaxToRemove
     * @return
     */
    @SuppressWarnings("boxing")
    public static double relativeDeviationTimeWithoutMinMax(final List<Long> executionTimes, final int numberOfMinMaxToRemove)
    {
        final long averageTimeWithoutMinMax = averageTimeWithoutMinMax(executionTimes, numberOfMinMaxToRemove);
        long deviation = 0;
        final int minMax = (numberOfMinMaxToRemove > 0) ? numberOfMinMaxToRemove : 0; 
        if (minMax != 0)
        {
            Collections.sort(executionTimes);
        }
        for (int i = minMax; i < executionTimes.size() - minMax; i++)
        {
            deviation += Math.pow(executionTimes.get(i) - averageTimeWithoutMinMax, 2);
        }
        return ((int) ((Math.sqrt(deviation / (executionTimes.size() - 2 * minMax)) / averageTimeWithoutMinMax * 100) * 100)) / 100d;
    }    

    /**
     * testFiboStatic.
     * 
     * @return
     */
    public static long testFiboStatic()
    {
        long start = System.currentTimeMillis();
        int fibo = 0;
        start = System.currentTimeMillis();
        for (int i = NB_RECURSIVE_TESTS; i-- != 0;)
        {
            fibo = FiboInstance.fibStatic(FIBO_N);
        }
        long end = System.currentTimeMillis();
        long executionTime = end - start;
        System.out.println("[RecursiveTest], Invoke of " + NB_RECURSIVE_TESTS + " int fibo = FiboInstance.fibStatic(" + FIBO_N + "), fibo=" + fibo + ", snapshot time,"
                + executionTime);
        return executionTime;
    }
    
    /**
     * testFibo2.
     * 
     * @return
     */
    public static long testFiboInstance()
    {
        long start = System.currentTimeMillis();
        int fibo = 0;
        final FiboInstance fiboInstance = new FiboInstance();
        start = System.currentTimeMillis();
        for (int i = NB_RECURSIVE_TESTS; i-- != 0;)
        {
            fibo = fiboInstance.fib(FIBO_N);
        }
        long end = System.currentTimeMillis();
        long executionTime = end - start;
        System.out.println("[RecursiveTest], Invoke of  " + NB_RECURSIVE_TESTS + " int fibo = fiboInstance.fib(" + FIBO_N + "), fibo=" + fibo + ", snapshot time,"
                + executionTime);
        return executionTime;
    }

    
    
    /**
     * testFiboInstanceImplWithInterface.
     * 
     * @return
     */
    public static long testFiboInstanceImplWithInterface()
    {
        long start = System.currentTimeMillis();
        int fiboResult = 0;
        final Fibo fibo = new FiboImpl();
        start = System.currentTimeMillis();
        for (int i = NB_RECURSIVE_TESTS; i-- != 0;)
        {
            fiboResult = fibo.fib(FIBO_N);
        }
        long end = System.currentTimeMillis();
        long executionTime = end - start;
        System.out.println("[RecursiveTest], Invoke of " + NB_RECURSIVE_TESTS + " fiboResult = fibo.fib(" + FIBO_N + "), fibo=" + fiboResult + ", snapshot time,"
                + executionTime);
        return executionTime;
    }
    

    
    /**
     * testFiboInstanceImplWithoutInterface.
     * 
     * @return
     */
    public static long testFiboInstanceImplWithoutInterface()
    {
        long start = System.currentTimeMillis();
        int fibo = 0;
        final FiboImpl fiboImpl = new FiboImpl();
        start = System.currentTimeMillis();
        for (int i = NB_RECURSIVE_TESTS; i-- != 0;)
        {
            fibo = fiboImpl.fib(FIBO_N);
        }
        long end = System.currentTimeMillis();
        long executionTime = end - start;
        System.out.println("[RecursiveTest], Invoke of " + NB_RECURSIVE_TESTS + " fibo = fiboImpl.fib(" + FIBO_N + "), fibo=" + fibo + ", snapshot time,"
                + executionTime);
        return executionTime;
    }
    
    

    /**
     * main.
     * 
     * @param args
     */
    @SuppressWarnings("boxing")
    public static void main(final String[] args)
    {
        final int nbTests = (args != null && args.length >= 1 && args[0] != null) ?  Integer.valueOf(args[0]): NB_TESTS;
        final int nbOfExclusionMinMax = (args != null && args.length >= 2 && args[1] != null) ? Integer.valueOf(args[1]) : NB_OF_EXCLUSION_MIN_MAX;
        
        final List<Long> executionTimes = new ArrayList<Long>(nbTests);
        
        for (int i = nbTests; i != 0; i--)
            executionTimes.add(RecursiveTest.testFiboStatic());
        System.out.println("[RecursiveTest], Invoke of " + NB_RECURSIVE_TESTS + " int fibo = FiboInstance.fibStatic(" + FIBO_N + "),, average time,"
                + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) + ", min time," + Collections.min(executionTimes) + ", max time," + Collections.max(executionTimes)
                + ", relative deviation time," + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));
        executionTimes.clear();
        
        for (int i = nbTests; i != 0; i--)
            executionTimes.add(RecursiveTest.testFiboInstance());
        System.out.println("[RecursiveTest], Invoke of " + NB_RECURSIVE_TESTS + " int fibo = fiboInstance.fib(" + FIBO_N + "),, average time,"
                + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) + ", min time," + Collections.min(executionTimes) + ", max time," + Collections.max(executionTimes)
                + ", relative deviation time," + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));
        executionTimes.clear();
        
        for (int i = nbTests; i != 0; i--)
            executionTimes.add(RecursiveTest.testFiboInstanceImplWithInterface());
        System.out.println("[RecursiveTest], Invoke of " + NB_RECURSIVE_TESTS + " fiboResult = fibo.fib(" + FIBO_N + "),, average time,"
                + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) + ", min time," + Collections.min(executionTimes) + ", max time," + Collections.max(executionTimes)
                + ", relative deviation time," + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));
        executionTimes.clear();
        
        for (int i = nbTests; i != 0; i--)
            executionTimes.add(RecursiveTest.testFiboInstanceImplWithoutInterface());
        System.out.println("[RecursiveTest], Invoke of " + NB_RECURSIVE_TESTS + " fibo = fiboImpl.fib(" + FIBO_N + "),, average time,"
                + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) + ", min time," + Collections.min(executionTimes) + ", max time," + Collections.max(executionTimes)
                + ", relative deviation time," + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));
        executionTimes.clear();
        
    }
    


}

/**
 * Fibo.
 * 
 * @author anavarro122404 - 25 sept. 07
 *
 *
 * <!-- $Id: RecursiveTest.java,v 1.2 2007/11/20 13:17:31 anavarro Exp $ -->.
 *
 */
final class FiboInstance
{

    /**
     * fib.
     *
     * @param n
     * @return
     */
    public final static int fibStatic(final int n)
    {
        if (n < 2)
        //if (n == 0 || n == 1)
            return (1);
        return (fibStatic(n - 2) + fibStatic(n - 1));
    }

    
    /**
     * fib.
     *
     * @param n
     * @return
     */
    public int fib(final int n)
    {
        if (n < 2)
        //if (n == 0 || n == 1)
            return (1);
        return (fib(n - 2) + fib(n - 1));
    }

}

/**
 * Fibonacci.
 * 
 * @author anavarro122404 - 7 Nov 2007
 *
 *
 * <!-- $Id: RecursiveTest.java,v 1.2 2007/11/20 13:17:31 anavarro Exp $ -->.
 *
 */
interface Fibo
{
    /**
     * fib.
     *
     * @param n
     * @return
     */
    public abstract int fib(final int n);
    
}

/**
 * FibonacciImpl.
 * 
 * @author anavarro122404 - 8 Nov 2007
 *
 *
 * <!-- $Id: RecursiveTest.java,v 1.2 2007/11/20 13:17:31 anavarro Exp $ -->.
 *
 */
final class FiboImpl implements Fibo
{
    /**
     * fib.
     *
     * @param n
     * @return
     */
    public int fib(final int n)
    {
        if (n < 2)
            return (1);
        return (fib(n - 2) + fib(n - 1));
    }

}

